language/stdlib/modules/libra.move

```diff
@@ -100,27 +102,28 @@ module Libra {
         T<Token> { value }
     }
 
-    // Create a new Preburn resource.
-    // Can only be called by the holder of the MintCapability.
-    public fun new_preburn_with_capability<Token>(
-        _capability: &MintCapability<Token>
-    ): Preburn<Token> {
-        assert_is_registered<Token>();
-        Preburn<Token> { requests: Vector::empty() }
-    }
-
-    // Send coin to the preburn holding area, where it will wait to be burned.
-    // Fails if the sender does not have a published Preburn resource
-    public fun preburn<Token>(coin: T<Token>) acquires Info, Preburn {
+    // Send coin to the preburn holding area `preburn_ref`, where it will wait to be burned.
+    public fun preburn<Token>(
+        preburn_ref: &mut Preburn<Token>,
+        coin: T<Token>
+    ) acquires Info {
+        // TODO: bring this back once we can automate approvals in testnet
+        // Transaction::assert(preburn_ref.is_approved, 13);
         let coin_value = value(&coin);
         Vector::push_back(
-            &mut borrow_global_mut<Preburn<Token>>(Transaction::sender()).requests,
+            &mut preburn_ref.requests,
             coin
         );
         let market_cap = borrow_global_mut<Info<Token>>(0xA550C18);
         market_cap.preburn_value = market_cap.preburn_value + coin_value
     }
 
+    // Send coin to the preburn holding area, where it will wait to be burned.
+    // Fails if the sender does not have a published Preburn resource
+    public fun preburn_from_sender<Token>(coin: T<Token>) acquires Info, Preburn {
```

Comment:
I'm a little puzzled by this function name. It is not taking anything *from* the sender. If anything, it seems like this should be called "preburn_to_sender". Maybe I don't understand the naming scheme. But, aside from that, why not just keep the one "preburn" function? I skimmed this PR to see if there are places where you want to call "preburn", but it looks like everything is now using the new name "preburn_from_sender". Do we really need both?

Reply:
Good questions. The naming convention we use else

>It is not taking anything from the sender. If anything, it seems like this should be called "preburn_to_sender". Maybe I don't understand the naming scheme.

I was modelling this after `LibraAccount::pay_from_sender`/`withdraw_from_sender`, where the `X_from_sender` means "this does `X` using a resource published under the sender's account". Definitely open to hearing different naming conventions though; agreed that the `from` sounds like it's talking about the directionality of the transfer.

>But, aside from that, why not just keep the one "preburn" function?
Doing it this way give us the flexibility to `Preburn` from both a `Preburn` resource published at the top level and a `Preburn` resource stored in a field of a different resource. We anticipate the need for the latter feature in a synthetic currency where burning/preburning is not a privileged operation (I believe @tzakian can show you some example code for this).

Reply:
Maybe "preburn_for_sender"?

Reply:
I am open to this, but could you propose a rule for when to use `for_sender` vs `from_sender` (or are you suggesting that we should also write `LibraAccount::pay_for_sender`)?

Reply:
I was hoping that you or @tzakian might have ideas about a general rule.... The other two functions (withdraw_from_sender and pay_from_sender) both make sense to me with "from_sender" because they are taking funds away "from" the sender's account. If you want a general rule, we could go with "to_sender" and "from_sender" based on whether the function is adding resources to or removing resources from the sender's account.

Reply:
The `to`/`from` discipline seems sensible to me; I like that they are clear opposites. There are some potentially odd cases here if a procedure both adds and removes (e.g., what would the procedure `x` in `LibraAccount` be called?)

```
x(amount: u64) {
   Libra::preburn_to_sender(withdraw_from_sender(amount))
}
```

But that seems uncommon enough not to be a problem. If it's ok, I'll rename to `preburn_to_sender` for now.